File: 06.py
from typing import List
from input_reader import readLines

def countLanternFish(startingFish: List[int], days: int) -> int:
    resetDayCount, initialDayCount = 6, 8
    daysRemaining = days
    fishTimers = [ sum([ x == i for x in startingFish ]) for i in range(0, initialDayCount+1) ]
    for _ in range(0, daysRemaining):
        fishTimers = [ fishTimers[i+1] if i != resetDayCount else fishTimers[i+1] + fishTimers[0] for i in range(0, initialDayCount) ] + [ fishTimers[0] ]
    return sum(fishTimers)

def part1(xs: List[int]) -> int:
    return countLanternFish(xs, 80)

def part2(xs: List[int]) -> int:
    return countLanternFish(xs, 256)

def main():
    xs = [ int(x) for x in readLines("06")[0].replace("\n", "").split(",") ]

    print("Part 1", part1(xs))
    print("Part 2", part2(xs))

if __name__ == "__main__":
    main()

File: 06.hs
import Split (split)

resetDayCount = 6
initialDayCount = 8

getInitialFishCount :: [Int] -> [Int]
getInitialFishCount xs = map (\x -> length $ filter (==x) xs) [0..initialDayCount]

countLanternFish :: Int -> [Int] -> Int
countLanternFish 0 xs    = sum xs
countLanternFish days xs = countLanternFish (days - 1) (map updateTimers [0..initialDayCount])
    where
        updateTimers :: Int -> Int
        updateTimers 6 = head xs + xs !! (resetDayCount+1)
        updateTimers 8 = head xs
        updateTimers x = xs !! (x+1)

f :: [Int] -> Int
f = countLanternFish 80 . getInitialFishCount

g :: [Int] -> Int
g = countLanternFish 256 . getInitialFishCount

main = do
    contents <- readFile "input/06.txt"
    let ss = lines contents
        xs :: [Int]
        xs = map read $ split "," $ head ss
    print ("#1: " ++ show (f xs))
    print ("#2: " ++ show (g xs))